File: app/src/main/java/com/starostinvlad/fan/SerialPageScreen/SerialPageScreenPresenter.java
package com.starostinvlad.fan.SerialPageScreen;

import android.util.Log;

import com.starostinvlad.fan.App;
import com.starostinvlad.fan.BaseMVP.BasePresenter;
import com.starostinvlad.fan.GsonModels.CurrentSerial;
import com.starostinvlad.fan.VideoScreen.PlayerModel.Serial;
import com.starostinvlad.fan.VideoScreen.PlayerModel.SerialPlayer;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

class SerialPageScreenPresenter extends BasePresenter {

    private final SerialPageScreenContract view;
    private final SerialPlayer serialPlayer;
    private Serial serial;
    private final String TAG = getClass().getSimpleName();

    SerialPageScreenPresenter(SerialPageScreenContract serialPageScreenContract) {
        this.view = serialPageScreenContract;
        serialPlayer = new SerialPlayer(App.getInstance().getDomain(), App.getInstance().getOkHttpClient());
    }

    void putToSubscribe(String id, boolean checked) {
        disposables.add(
                serialPlayer.subscribeRequest(id, checked)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                val -> Log.d(TAG, "answer: " + val),
                                throwable -> {
//                                    sendErrorMsg(throwable.getMessage());
                                    throwable.printStackTrace();
                                }
                        )
        );
    }

    void loadData(String url) {
        view.showLoading(true);

        disposables.add(
                serialPlayer
                        .getSerial(url)
                        .flatMap(
                                val -> App.getInstance()
                                        .getDatabase()
                                        .currentSerialDao()
                                        .getById(val.getId())
                                        .defaultIfEmpty(new CurrentSerial(
                                                val.getId(),
                                                val.getCurrentSeasonIndex(),
                                                val.getCurrentEpisodeIndex(),
                                                val.getCurrentTranslationIndex()
                                        )),
                                ((serial1, currentSerial) -> {
                                    if (currentSerial != null) {
                                        serial1.setCurrentSeasonIndex(currentSerial.getCurrentSeasonIndex());
                                        serial1.setCurrentEpisodeIndex(currentSerial.getCurrentEpisodeIndex());
                                        serial1.setCurrentTranslationIndex(currentSerial.getCurrentTranslationIndex());
                                    }
                                    return serial1;
                                })
                        )
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(serial -> {
                            this.serial = serial;

                            view.fillBtn(
                                    serial.getCurrentSeasonIndex() + 1,
                                    serial.getCurrentEpisodeIndex() + 1
                            );

                            view.fillPage(serialPlayer);
                            view.fillSeasonsList(serial);

                            view.checkSubscribed(serialPlayer.isSubscribed());

                            Log.d(TAG, "loadData: serial: " + serial);
                            view.showLoading(false);
                        }, Throwable::printStackTrace)
        );
    }

    void openSerialOnClick() {
        view.openActivityWithSerial(serial);
    }
}

File: app/src/main/java/com/starostinvlad/fan/SerialPageScreen/SerialPageScreenPresenter.kt
package com.starostinvlad.fan.SerialPageScreen

import android.util.Log
import com.starostinvlad.fan.App
import com.starostinvlad.fan.BaseMVP.BasePresenter
import com.starostinvlad.fan.GsonModels.CurrentSerial
import com.starostinvlad.fan.VideoScreen.PlayerModel.Serial
import com.starostinvlad.fan.VideoScreen.PlayerModel.SerialPlayer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

internal class SerialPageScreenPresenter : BasePresenter<SerialPageScreenContract?>() {
    private var serialPlayer: SerialPlayer? = null
    private var serial: Serial? = null
    private val TAG: String = this::class.simpleName!!
    override fun attachView(mvpView: SerialPageScreenContract?) {
        super.attachView(mvpView)
        serialPlayer = SerialPlayer(App.instance.domain, App.instance.okHttpClient)
    }

    fun putToSubscribe(id: String?, checked: Boolean) {
        disposables.add(
                serialPlayer!!.subscribeRequest(id, checked)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(
                                { `val`: String? -> Log.d(TAG, "answer: $`val`") }
                        ) { throwable: Throwable ->
//                                    sendErrorMsg(throwable.getMessage());
                            throwable.printStackTrace()
                        }
        )
    }

    fun loadData(url: String) {
        view!!.showLoading(true)
        disposables.add(
                serialPlayer!!
                        .getSerial(url)
                        .flatMap(
                                { serial ->
                                    App.instance
                                            .database
                                            .currentSerialDao()
                                            .getById(serial.id)
                                            .defaultIfEmpty(CurrentSerial(
                                                    serial.id,
                                                    serial.currentSeasonIndex,
                                                    serial.currentEpisodeIndex,
                                                    serial.currentTranslationIndex
                                            ))
                                }
                        ) { serial1: Serial?, currentSerial: CurrentSerial? ->
                            if (currentSerial != null) {
                                serial1?.currentSeasonIndex = currentSerial.currentSeasonIndex
                                serial1?.currentEpisodeIndex = currentSerial.currentEpisodeIndex
                                serial1?.currentTranslationIndex = currentSerial.currentTranslationIndex
                            }
                            serial1!!
                        }
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({ serial: Serial? ->
                            this.serial = serial
                            view!!.fillBtn(
                                    serial!!.currentSeasonIndex + 1,
                                    serial.currentEpisodeIndex + 1
                            )
                            view!!.fillPage(serialPlayer)
                            view!!.fillSeasonsList(serial)
                            view!!.checkSubscribed(serialPlayer!!.isSubscribed)
                            Log.d(TAG, "loadData: serial: $serial")
                            view!!.showLoading(false)
                        }) { obj: Throwable -> obj.printStackTrace() }
        )
    }

    fun openSerialOnClick() {
        view!!.openActivityWithSerial(serial)
    }
}